### PR TITLE
Allow projectiles to emit sub-projectiles while they are alive

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature**

This PR implements a feature I thought about a while back and wanted to wait until projectile penetration was a thing before doing.

## Summary
Currently, weapons are able to spawn projectiles that split into smaller projectiles when the main projectile dies. What it is impossible to do though is have a projectile spawn smaller projectiles over the course of its lifetime.

This PR introduces an "emission" mechanic. Projectiles are able to have various emitters that can spawn sub projectiles over the course of the main projectile's lifetime. This adds the following weapon attribute:
* `"emission" <weapon name> [<projectile count>]`: This weapon's projectiles will emit projectiles from the weapon of the given name. If a projectile count is given, then that number of projectiles will spawn every time that the main projectile emits. If not listed, then only one projectile is spawned per emission. The `"emission"` keyword can have the following children:
  * `"facing" <angle>`: Behaves like "facing" on a submunition, changing the angle that the subprojectile is spawned at relative to the main projectile.
  * `"offset" <x> <y>`: Behaves like "offset" on a submunition, changing the location that the subprojectile is spawned at relative to the main projectile.
  * `"arming time" <frames>`: If present, this many frames must pass before the main projectile can start emitting subprojectiles. If not present, the main projectile will immediately fire its subprojectiles upon being spawned in (assuming that other conditions for emitting are met).
  * `"emission count" <count>`: If present, the main projectile can only emit up to this many times over the course of its lifetime. After the emission count is reached, no more subprojectiles will be emitted. If not listed, the only limit is the lifetime of the projectile.
  * `"emission range" <distance>`: If present, the main projectile must have a target in order to emit its subprojectiles. The target must be within this range for an emission to occur.
  * `"aim at target"`: A valueless attribute that if present causes subprojectiles to spawn in facing the target of the main projectile. This does not prevent subprojectiles from emitting if there is no target.

The rate at which emissions occur is controlled by the reload, burst reload, and burst count attributes of the emission weapon. So for example, if a weapon has a lifetime of 100 frames, an emission with an arming time of 20 frames, and the emission weapon has a reload of 10 frames, then the projectile will take 20 frames to arm, then spawn in subprojectiles from the emission every 10 frames until the main projectile dies.

Unlike submunitions, the emissions that a projectile has do not increase the damage of the projectile when it impacts a ship, as submunitions will add their damage to the damage of the main projectile upon impact if the submunitions never got a chance to spawn in.

## Usage examples
The following "Mini Missile Spawner" weapon would spawn a projectile that deals 100 shield or hull damage upon impact. After 10 frames, the spawner will be armed and start spawning two mini-missiles at a time. These mini-missiles will fire in bursts of 3 each with 2 frames in between until waiting for 10 frames per missile in the burst to fire again. Since the spawner has an emission count of 2, this will result in waves of 6 mini missiles. Since subprojectiles share the target of the main projectile, these subprojectiles will home in on whatever ship the spawner was targetting, although the spawner itself has no tracking capabilities/
```
outfit "Mini Missile Spawner"
	weapon
		sprite "projectile/mini missile spawn"
		emission "Mini Missile" 2
			"arming time" 10
			"aim at target"
			"emission range" 500
		lifetime 100
		velocity 5
		"shield damage" 100
		"hull damage" 100
		"missile strength" 100

outfit "Mini Missile"
	weapon
		inaccuracy 20
		sprite "projectile/mini missile"
		reload 10
		"burst reload" 2
		"burst count" 3
		"shield damage" 10
		"hull damage" 10
		"homing" 3
		"infrared tracking" .5
		lifetime 50
		velocity 10
		"missile strength" 5
```

## Testing Done
I'll get to it.

### CODE REVIEW

File: source/Projectile.cpp [hunk 1]
```diff
@@ -135,9 +148,47 @@ void Projectile::Move(vector<Visual> &visuals, vector<Projectile> &projectiles)
...
+	{
+		const Weapon::Emission &sourceEmission = emission.sourceEmission;
+		const Weapon *const subWeapon = sourceEmission.weapon;
+		if(sourceEmission.emissionCount && emission.fired >= sourceEmission.emissionCount)
+			continue;
+
+		double range = sourceEmission.emissionRange;
+		if(emission.armingTime)
+			--emission.armingTime;
+		else if(emission.burstReload <= 0. && emission.burstCount
```
Comment (Amazinite) [inline source/Projectile.cpp:162]:
```suggestion
		else if(emission.burstReload <= 0. && emission.burstCount
```

**Comment by Amazinite** [comment]
I'm open to renaming this. I could never think of a better name.

**Comment by Saugia** [comment]
I'd second a rename. Emitter is more of a phrase used for particle effects (at least in 3d environments), so when I read emitter, I think of a live effect as opposed to the above.

**Comment by Amazinite** [comment]
Perhaps these could also be named "submunition", but we have some child node that notes that the submunition should be created during the lifetime of the projectile instead of when it dies.

**Comment by Hurleveur** [comment]
Maybe with a word in front of submunition; like
`"live/active/permanent submunition"`?

**Comment by Amazinite** [comment]
I'll go with "live submunition" because it matches "live effect", which are effects created while the projectile is alive.

**Comment by brendanjones** [comment]
> Perhaps these could also be named "submunition", but we have some child node that notes that the submunition should be created during the lifetime of the projectile instead of when it dies.

Don't mind me coming into a random PR, I was just browsing through the PR list to see what's up. Is the correct term for a weapon that splits after launch not [Cluster Munition](https://en.wikipedia.org/wiki/Cluster_munition)? Also, I freakin love this idea, good stuff @Amazinite . Some very cool weapons could come from this. It's reminding me of the [SAE Launcher ](https://evn.fandom.com/wiki/Escape_Velocity_Override#Igadzra) in Override (though iirc they launched already split. Still, a cool weapon effect).

**Comment by tibetiroka** [comment]
@Amazinite Can you fix the merge conflicts?

**Comment by tibetiroka** [comment]
Closing as orphaned. Feel free to reopen this PR once the conflicts are resolved.

**Pull request closed without merging**